File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Stack<Character> stack = new Stack<>();

        for(char c : s.toCharArray()){
            if(c == '('){
                stack.push(')');
            } else if (c == '['){
                stack.push(']');
            } else if (c == '{'){
                stack.push('}');                
            } else {
                if(stack.isEmpty()){
                    return false;
                }
                char value = stack.pop();
                if(c != value && c != value && c != value){
                    return false;
                }
            }
        }
        
        return stack.isEmpty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        # Time : O(N)
        # Space : O(N)

        stack = []
        
        for c in s:
            if c == "(":
                stack.append(")")
            elif c == "[":
                stack.append("]")
            elif c == "{":
                stack.append("}")
            else:
                if not stack:
                    return False                
                if stack[-1] != c:
                    return False
                stack.pop()

        return True if not stack else False
